nvflare/app_opt/lightning/fl_callbacks.py

```diff
@@ -0,0 +1,157 @@
+            self.prev_loop_run = loop.run
+            loop.run = test_loop_run_decorator(loop, self)
+
+    def on_fit_start(self, trainer, pl_module):
+        # receive the global model and update the local model with global model
+        # the 1st time test() or fit() is called.
+        self._receive_update_model(pl_module)
+
+    def on_train_end(self, trainer, pl_module):
+        if self.output_fl_model:
+            self.output_fl_model.params = pl_module.cpu().state_dict()
+        else:
+            self.output_fl_model = flare.FLModel(params=pl_module.cpu().state_dict())
+        self._check_and_send()
+
+    def on_test_start(self, trainer, pl_module):
+        # receive the global model and update the local model with global model
+        # the 1st time test() or train() is called.
+        # expect user will validate the global model first (i.e. test()), once that's done.
+        # the metrics_captured will be set to True.
+        # The subsequence test() calls will not trigger the receive update model.
+        # Hence the test() will be validating the local model.
+        if pl_module and not self.metrics_captured:
+            self._receive_update_model(pl_module)
+
+    def on_test_end(self, trainer, pl_module):
+        self._check_and_send()
+
+    def teardown(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: str) -> None:
+        if self.prev_loop_run:
+            trainer.test_loop.run = self.prev_loop_run
+
+    def _receive_update_model(self, pl_module):
+        if not self.input_fl_model:
+            model = self._receive_model()
+            if model and model.params:
+                pl_module.load_state_dict(model.params)
+
+    def _receive_model(self) -> FLModel:
+        model = flare.receive()
+        if model:
+            self.input_fl_model = model
+        return model
+
+    def _check_and_send(self):
+        if self.output_fl_model:
+            if self.send_mode == SendTrigger.AFTER_TRAIN_AND_TEST:
+                if self.output_fl_model.metrics and self.output_fl_model.params:
+                    self.send()
+            elif self.send_mode == SendTrigger.AFTER_TRAIN and self.output_fl_model.params:
+                self.send()
+            elif self.send_mode == SendTrigger.AFTER_TEST and self.output_fl_model.metrics:
+                self.send()
+
+    def send(self):
+        if not self.model_sent:
+            try:
+                flare.send(self.output_fl_model)
+                self.model_sent = True
+            except Exception as e:
+                raise RuntimeError("failed to send FL model", e)
+
+
+def test_loop_run_decorator(loop, cb):
+    func = loop.run
+
+    def wrapper(*args, **kwargs):
+        if cb.metrics_captured or cb.send_mode == SendTrigger.AFTER_TRAIN:
+            try:
+                return func(*args, **kwargs)
+            except BaseException as e:
+                print(traceback.format_exc())
+                raise e
```

Comment:
I think we should just `return func(*args, **kwargs)` here without try except.

Users can do try except block on the outside when calling trainer.test

Reply:
If we doing this here, we should also be doing this try-except block for the else part to be consistent?